diff --git a/src/BODY_F_diag_withref.hf b/src/BODY_F_diag_withref.hf
index 6523ac0..9360f9e 100644
--- a/src/BODY_F_diag_withref.hf
+++ b/src/BODY_F_diag_withref.hf
@@ -1,3 +1,4 @@
+#define is_contiguous(X) .false.
       use vgrid_utils, only: get_allocate
       ! Given referent, compute physical levelling information from the vertical description
       type(vgrid_descriptor), intent(in) :: self                  !Vertical descriptor instance
diff --git a/src/Makefile b/src/Makefile
index c4b5cad..8e2d025 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -1,14 +1,4 @@
 
-ifeq ($(BASE_ARCH),$(EC_ARCH))
-   $(info ERROR: please setup the compilation environment by typing the folowing:)
-   $(info . setup.dot or . setup_gpsc.dot)
-   $(error )
-endif
-
-ifeq ($(USE_EFENCE),1)
-   LEFENCE=-lefence
-endif
-
 include version.dot
 
 default: libdescrip.a
@@ -17,12 +7,8 @@ dependencies.mk: *.c *.F90 *.h *.hc
 	ls *.hc *.c *.F90 | ./s.dependencies.pl > dependencies.mk ;\
 
 include dependencies.mk
-
-%.o: %.F90
-	s.f90 -c -openmp $(MY_FFLAGS) $(FFLAGS) $(LEFENCE) $<
-
-%.o: %.c
-	s.cc -c -openmp $(MY_CFLAGS) $(CFLAGS) $(LEFENCE) $<
+.SUFFIXES : .c .F90 .o
+include $(PROJECT_ROOT)/include/makefile_suffix_rules.inc
 
 libdescrip.a: $(OBJECTS) dependencies.mk vgrid_version
 	rm -f libdescrip.a && ar scru libdescrip.a *.o ;\
diff --git a/src/dependencies.mk b/src/dependencies.mk
new file mode 100644
index 0000000..12cbd65
--- /dev/null
+++ b/src/dependencies.mk
@@ -0,0 +1,33 @@
+INCDECKS=
+F90DECKS=
+FTN90DECKS=
+TMPL90DECKS=
+PTNDECKS=
+CDECKS= \
+	vgrid.c  
+FHDECKS=
+CDKDECKS=
+CHDECKS=
+HFDECKS= \
+	BODY_F_diag_withref.hf  BODY_F_diag_withref_prof.hf  BODY_F_dpidpis_withref.hf  BODY_F_dpidpis_withref_prof.hf  BODY_F_levels_withref.hf   \
+	BODY_F_levels_withref_prof.hf  vgrid_descriptors.hf  
+FTNDECKS=
+HDECKS= \
+	armnlib.h  vgrid.h  
+HCDECKS= \
+	BODY_C_compute_heights_0001.hc  BODY_C_compute_heights_21001.hc  BODY_C_compute_heights_4001.hc  BODY_C_compute_pressure_1001_1002.hc  BODY_C_compute_pressure_1003_5001.hc   \
+	BODY_C_compute_pressure_2001.hc  BODY_C_compute_pressure_5002_5003_5004_5005.hc  BODY_C_compute_pressure_5100.hc  BODY_Cvgd_diag_withref_2ref.hc  
+FDECKS=
+CDK90DECKS=
+PTN90DECKS=
+OBJECTS= \
+	vgrid.o  vgrid_descriptors.o  vgrid_utils.o  
+vgrid.o:	vgrid.c \
+	BODY_C_compute_pressure_1003_5001.hc  vgrid.h  armnlib.h  BODY_Cvgd_diag_withref_2ref.hc  BODY_C_compute_pressure_5100.hc   \
+	BODY_C_compute_heights_4001.hc  BODY_C_compute_heights_21001.hc  BODY_C_compute_pressure_5002_5003_5004_5005.hc  BODY_C_compute_pressure_2001.hc  BODY_C_compute_heights_0001.hc   \
+	BODY_C_compute_pressure_1001_1002.hc  
+vgrid_descriptors.o:	vgrid_descriptors.F90 \
+	BODY_F_diag_withref.hf  vgrid_utils.o  BODY_F_levels_withref.hf  BODY_F_dpidpis_withref_prof.hf  BODY_F_diag_withref_prof.hf   \
+	BODY_F_levels_withref_prof.hf  vgrid_descriptors.hf  BODY_F_dpidpis_withref.hf  
+vgrid_utils.o:	vgrid_utils.F90 \
+	vgrid_descriptors.hf  
